octoprint_firmwareupdate/__init__.py

```diff
@@ -1,85 +1,271 @@
+            self._logger.info("File downloaded, continuing...")
+            self._update_firmware("github")
+        else:
+            self._update_status(False, "error", "Release firmware was not downloaded.")
+
+    def _update_firmware(self, target):
+        if not self.isUpdating:
+            # TODO: Update this error message to be more helpful
+            self._logger.info("Something doesn't make sense here...")
+        else:
+            self._update_firmware_thread = Thread(target=self._update_worker)
+            self._update_firmware_thread.daemon = True
+            self._update_firmware_thread.start()
+
+    def _update_worker(self):
+        self._logger.info("Updating now...")
+
+        try:
+            self.port = glob('/dev/ttyACM*')[0]
+        except IndexError:
+            self._update_status(False, "error", "No ports exist.")
+            raise RuntimeError('No ports detected')
+
+        try:
+            os.remove(os.path.join(os.path.expanduser('~'), 'Marlin/.build_log'))
+        except OSError:
+            self._logger.info("Couldn't delete build log file")
+        self.f = open(os.path.join(os.path.expanduser('~'), 'Marlin/.build_log'), "w")
+        try:
+            s = Serial(self.port, 115200)
+        except SerialException as e:
+            self._update_status(False, "error", str(e))
+            raise RuntimeError(str(e))
+        s.setDTR(False)
+        sleep(0.1)
+        s.setDTR(True)
+        s.close()
+        pro = Popen("cd ~/Marlin/; avrdude -p m2560 -P /dev/ttyACM0 -c stk500v2 -b 250000 -D -U flash:w:./.build/mega2560/firmware.hex:i", stdout=self.f, stderr=self.f, shell=True, preexec_fn=os.setsid)
+        self.updatePID = pro.pid
+        self.checkStatus()
+
+    def find_between(self, s, first, last):
+        try:
+            start = s.rindex(first) + len(first)
+            end = s.rindex(last, start)
+            return s[start:end]
+        except ValueError:
+            return ""
+
+    def _update_status(self, isUpdating, status=None, message=None):
+        self.isUpdating = isUpdating
+        # Reconnect again after no longer updating
+        if not self.isUpdating:
+            self._printer.connect()
+            if status == "error":
+                try:
+                    os.remove(self.version_file)
+                except OSError:
+                    pass
+        else:
+            self._printer.disconnect()
+
+        self._plugin_manager.send_plugin_message(self._identifier, dict(isUpdating=self.isUpdating, status=status, message=message))
+        payload = {'isUpdating': self.isUpdating, 'status': status, 'message': message}
+        eventManager().fire(Events.FIRMWARE_UPDATE, payload)
+
+    def _clean_up(self):
+        if self.f is not None:
+            if not self.f.closed:
+                self.f.close()
+        try:
+            os.remove(os.remove(os.path.join(os.path.expanduser('~'), 'Marlin/.build_log')))
```

Comment:
Maybe we should leave behind the build log and firmware file for forensics incase we need to figure out what happened.